README.md
# isodatetime.py

A bunch of helper functions for working with iso dates / datetimes in Python.

At the moment there is no "packaging" release, but the module (`isodatetime.py`) is a single Python file with no external dependencies, so you can just copy/paste it into your own projects,

## Demo
You can demo the functionality by running `demo.py` (which _does_ have some external dependencies).

![](docs/img/isodatetime_demo.png)

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 1.0.1 - 2023-01-20
### Added
- Add quotes around string return values in demo

## 1.0.0 - 2023-01-20
Initial Release

isodatetime.py
from datetime import datetime

VERSION = "1.0.1"

LOGURU_FORMAT_ISO = "{time:YYYY-MM-DD}T{time:HH:mm:ss.SSSZZ} | {level:<8} | {message}"
LOGURU_FORMAT_ISO_UTC = "{time:YYYY-MM-DD!UTC}T{time:HH:mm:ss.SSS!UTC}Z | {level:<8} | {message}"


def datetime_to_iso_datetime(datetime_object, tz=None, microseconds=False):
    """Convert datetime.datetime object to ISO datetime of form YYYY-MM-DDTHH:MM:SS[+/-]FF:FF.

    Args:
        tz (tzfile or datetime.timezone):
            A  tzfile object (such as returned dateutil.tz.gettz())

            OR

            A datetime.timezone object (such as returned by 
                datetime.timezone() or datetime.timezone.utc())

        microseconds(bool):
            If True, then microseconds will be included in the returned
            iso time string (YYYY-MM-DDTHH:MM:SS.MMMMMM[+/-]FF:FF)

    If tz is None (default), then the current system timezone will be used.
    """
    dt = datetime_object.astimezone(tz)
    if not microseconds:
        dt = dt.replace(microsecond=0)
    return dt.isoformat()


def datetime_to_iso_date(datetime_object):
    """Convert datetime.datetime object to ISO date of form YYYY-MM-DD."""
    return datetime_object.strftime(r"%Y-%m-%d")


def iso_date_to_datetime(iso_date):
    """Convert ISO date of form YYYY-MM-DD to datetime.datetime object."""
    return datetime.strptime(iso_date, r"%Y-%m-%d")


def iso_date_to_date(iso_date):
    """Convert ISO date of form YYYY-MM-DD to datetime.date object."""
    return datetime.strptime(iso_date, r"%Y-%m-%d").date()


def iso_datetime_now(tz=None, microseconds=False):
    """Return the current date and time as an iso datetime.

    (For options, see datetime_to_iso_datetime)
    """
    return datetime_to_iso_datetime(datetime.now(), tz=tz, microseconds=microseconds)


def iso_date_now(tz=None):
    """Return the current date as an iso date.
    
    (For options, see datetime_to_iso_datetime)
    """
    return datetime_to_iso_datetime(datetime.now(), tz=tz)[:10]


def iso_datetime_to_filesafe(iso_datetime):
    """Convert an iso datetime to a filename-safe string.

    The returned string will contain only characters suitable for inclusion in filename
        "2023-01-20T10:53:59+08:00" ->
        "2023-01-20T10_53_59P08_00"
    """
    return iso_datetime.replace(':', '_').replace('+', 'P')


demo.py
#!/usr/bin/env python

# Standard Library
import datetime
from dateutil import tz
import sys

# Library
from rich.console import Console
from rich.theme import Theme
from loguru import logger

# Local
from isodatetime import (
    VERSION,
    datetime_to_iso_datetime,
    datetime_to_iso_date,
    iso_date_to_datetime,
    iso_date_to_date,
    iso_datetime_now,
    iso_date_now,
    iso_datetime_to_filesafe,
    LOGURU_FORMAT_ISO,
    LOGURU_FORMAT_ISO_UTC
)
# --------------------
# Rich output console
# --------------------
# fmt: off
THEME = Theme({
    "expression": "#00ff00",
    "text": "#ffff00",
    "object": "#00ffff",
})
# fmt: on
CONSOLE = Console(highlight=False, color_system='256', theme=THEME)
CONSOLE_PPRINT = Console(highlight=True, color_system='256', theme=THEME)
print_native = print
print = CONSOLE.print
pprint = CONSOLE_PPRINT.print


# fmt:off
FUNCTION_EXAMPLES = [
    # -------------------------------
    # datetime_to_iso_datetime()
    # -------------------------------
    {
        'heading': "datetime_to_iso_datetime()",
        'expression': "datetime_to_iso_datetime(datetime.datetime(2022, 5, 6, 12, 34, 56))",
    },
    {
        'expression': "datetime_to_iso_datetime(datetime.datetime(2022, 5, 6, 12, 34, 56), tz=tz.gettz('UTC'))",
    },
    {
        'expression': "datetime_to_iso_datetime(datetime.datetime(2022, 5, 6, 12, 34, 56, 9876), microseconds=True)",
    },
    # -------------------------------
    # datetime_to_iso_date()
    # -------------------------------
    {
        'heading': "datetime_to_iso_date()",
        'expression': "datetime_to_iso_date(datetime.datetime(2022, 5, 6, 12, 34, 56))",
    },
    # -------------------------------
    # iso_date_to_datetime()
    # -------------------------------
    {
        'heading': "iso_date_to_datetime()",
        'expression': "iso_date_to_datetime('2022-05-06')",
    },
    # -------------------------------
    # iso_date_to_date()
    # -------------------------------
    {
        'heading': "iso_date_to_date()",
        'expression': "iso_date_to_date('2022-05-06')",
    },
    # -------------------------------
    # iso_datetime_now()
    # -------------------------------
    {
        'heading': "iso_datetime_now()",
        'expression': "iso_datetime_now()",
    },
    {
        'expression': "iso_datetime_now(tz=tz.gettz('UTC'))",
    },
    {
        'expression': "iso_datetime_now(microseconds=True)",
    },
    # -------------------------------
    # iso_date_now()
    # -------------------------------
    {
        'heading': "iso_date_now()",
        'expression': "iso_date_now()",
    },
    {
        'expression': "iso_date_now(tz=tz.gettz('Asia/Tokyo'))",
    },
    # -------------------------------
    # iso_datetime_to_filesafe()
    # -------------------------------
    {
        'heading': "iso_datetime_to_filesafe()",
        'expression': "iso_datetime_to_filesafe('2023-01-20T10:53:59-08:00')",
    },
    {
        'expression': "iso_datetime_to_filesafe('2023-01-20T10:53:59+08:00')",
    },
]
# fmt:on

# fmt:off
LOGURU_EXAMPLES = [
    {
        'heading': 'loguru logging with format: LOGURU_FORMAT_ISO',
        'format': LOGURU_FORMAT_ISO
    },
    {
        'heading': 'loguru logging with format: LOGURU_FORMAT_ISO_UTC',
        'format': LOGURU_FORMAT_ISO_UTC
    }
]
# fmt:off
def main():
    print(f'Demoing version {VERSION} of isodatetime.py:')
    for case in FUNCTION_EXAMPLES:
        show_heading(case)
        expression = case['expression']
        print(f'   [expression]{expression}[/expression]')
        result = eval(expression)
        if isinstance(result, str):
            result = f'[text]"{result}"[/text]'
        else:
            result = f'[object]{result.__repr__()}[/object]'
        print(f'      {result}')

    for case in LOGURU_EXAMPLES:
        show_heading(case)
        logger.remove()
        logger.add(
            sys.stdout,
            format=case['format']
        )
        logger.info('(example log message)')

def show_heading(case):
    heading = case.get('heading')
    if heading:
        print('-' * 80)
        print(f'{heading}')


if __name__ == "__main__":
    main()
